README.md
# stac-fastapi-caching
Stac-fastapi built on Tile38 to support caching. This code is built on top of stac-fastapi-elasticsearch 0.1.0 with pyle38, a Python client for tile38. Tile38 itself is built on top of Redis to support geospatial queries. 

### References

- tile38 - https://tile38.com/
- pyle38 - https://github.com/iwpnd/pyle38
- stac-fastapi-elasticsearch - https://github.com/stac-utils/stac-fastapi-elasticsearch

### Important

This code is not production-ready. CRUD routes work with items and collections. Bbox queries should work, and point and polygon intersection. Searching lists of item and collection ids is also functional.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Deprecated

### Added

### Fixed

### Changed


stac_fastapi/caching/tests/conftest.py
import asyncio
import copy
import json
import os
from typing import Any, Callable, Dict, Optional

import pytest
import pytest_asyncio
from httpx import AsyncClient

from stac_fastapi.api.app import StacApi
from stac_fastapi.api.models import create_request_model
from stac_fastapi.caching.config import AsyncTile38Settings
from stac_fastapi.caching.core import (
    BulkTransactionsClient,
    CoreClient,
    TransactionsClient,
)
from stac_fastapi.caching.database_logic import COLLECTIONS_INDEX, ITEMS_INDEX
from stac_fastapi.caching.extensions import QueryExtension

# from stac_fastapi.caching.indexes import IndexesClient
from stac_fastapi.extensions.core import (  # FieldsExtension,
    ContextExtension,
    SortExtension,
    TokenPaginationExtension,
    TransactionExtension,
)
from stac_fastapi.types.config import Settings
from stac_fastapi.types.search import BaseSearchGetRequest, BaseSearchPostRequest

DATA_DIR = os.path.join(os.path.dirname(__file__), "data")


class Context:
    def __init__(self, item, collection):
        self.item = item
        self.collection = collection


class MockRequest:
    base_url = "http://test-server"

    def __init__(
        self, method: str = "GET", url: str = "XXXX", app: Optional[Any] = None
    ):
        self.method = method
        self.url = url
        self.app = app


class TestSettings(AsyncTile38Settings):
    class Config:
        env_file = ".env.test"


settings = TestSettings()
Settings.set(settings)


@pytest.fixture(scope="session")
def event_loop():
    return asyncio.get_event_loop()


def _load_file(filename: str) -> Dict:
    with open(os.path.join(DATA_DIR, filename)) as file:
        return json.load(file)


_test_item_prototype = _load_file("test_item.json")
_test_collection_prototype = _load_file("test_collection.json")


@pytest.fixture
def load_test_data() -> Callable[[str], Dict]:
    return _load_file


@pytest.fixture
def test_item() -> Dict:
    return copy.deepcopy(_test_item_prototype)


@pytest.fixture
def test_collection() -> Dict:
    return copy.deepcopy(_test_collection_prototype)


async def create_collection(txn_client: TransactionsClient, collection: Dict) -> None:
    await txn_client.create_collection(
        dict(collection), request=MockRequest, refresh=True
    )


async def create_item(txn_client: TransactionsClient, item: Dict) -> None:
    await txn_client.create_item(item, request=MockRequest, refresh=True)


async def delete_collections_and_items(txn_client: TransactionsClient) -> None:
    await refresh_indices(txn_client)
    await txn_client.database.delete_items()
    await txn_client.database.delete_collections()


async def refresh_indices(txn_client: TransactionsClient) -> None:
    try:
        await txn_client.database.client.indices.refresh(index=ITEMS_INDEX)
    except Exception:
        pass

    try:
        await txn_client.database.client.indices.refresh(index=COLLECTIONS_INDEX)
    except Exception:
        pass


@pytest_asyncio.fixture()
async def ctx(txn_client: TransactionsClient, test_collection, test_item):
    # todo remove one of these when all methods use it
    await delete_collections_and_items(txn_client)

    try:
        await create_collection(txn_client, test_collection)
    except Exception:
        pass
    await create_item(txn_client, test_item)

    yield Context(item=test_item, collection=test_collection)

    await delete_collections_and_items(txn_client)


@pytest.fixture
def core_client():
    return CoreClient(session=None)


@pytest.fixture
def txn_client():
    return TransactionsClient(session=None)


@pytest.fixture
def bulk_txn_client():
    return BulkTransactionsClient(session=None)


@pytest_asyncio.fixture(scope="session")
async def app():
    settings = AsyncTile38Settings()
    extensions = [
        TransactionExtension(
            client=TransactionsClient(session=None), settings=settings
        ),
        ContextExtension(),
        SortExtension(),
        # FieldsExtension(),
        QueryExtension(),
        TokenPaginationExtension(),
    ]

    get_request_model = create_request_model(
        "SearchGetRequest",
        base_model=BaseSearchGetRequest,
        extensions=extensions,
        request_type="GET",
    )

    post_request_model = create_request_model(
        "SearchPostRequest",
        base_model=BaseSearchPostRequest,
        extensions=extensions,
        request_type="POST",
    )

    return StacApi(
        settings=settings,
        client=CoreClient(
            session=None,
            extensions=extensions,
            post_request_model=post_request_model,
        ),
        extensions=extensions,
        search_get_request_model=get_request_model,
        search_post_request_model=post_request_model,
    ).app


@pytest_asyncio.fixture(scope="session")
async def app_client(app):
    # await IndexesClient().create_indexes()

    async with AsyncClient(app=app, base_url="http://test-server") as c:
        yield c


stac_fastapi/caching/tests/api/test_api.py
import copy
import uuid
from datetime import datetime, timedelta

import pytest

from ..conftest import MockRequest, create_collection, create_item

ROUTES = {
    "GET /_mgmt/ping",
    "GET /docs/oauth2-redirect",
    "HEAD /docs/oauth2-redirect",
    "GET /",
    "GET /conformance",
    "GET /api",
    "GET /api.html",
    "HEAD /api",
    "HEAD /api.html",
    "GET /collections",
    "GET /collections/{collection_id}",
    "GET /collections/{collection_id}/items",
    "GET /collections/{collection_id}/items/{item_id}",
    "GET /search",
    "POST /search",
    "DELETE /collections/{collection_id}",
    "DELETE /collections/{collection_id}/items/{item_id}",
    "POST /collections",
    "POST /collections/{collection_id}/items",
    "PUT /collections",
    "PUT /collections/{collection_id}/items",
}


async def test_post_search_content_type(app_client, ctx):
    params = {"limit": 1}
    resp = await app_client.post("/search", json=params)
    assert resp.headers["content-type"] == "application/geo+json"


async def test_get_search_content_type(app_client, ctx):
    resp = await app_client.get("/search")
    assert resp.headers["content-type"] == "application/geo+json"


async def test_api_headers(app_client):
    resp = await app_client.get("/api")
    assert (
        resp.headers["content-type"] == "application/vnd.oai.openapi+json;version=3.0"
    )
    assert resp.status_code == 200


async def test_router(app):
    api_routes = set([f"{list(route.methods)[0]} {route.path}" for route in app.routes])
    assert len(api_routes - ROUTES) == 0


async def test_app_transaction_extension(app_client, ctx):
    item = copy.deepcopy(ctx.item)
    item["id"] = str(uuid.uuid4())
    resp = await app_client.post(f"/collections/{item['collection']}/items", json=item)
    assert resp.status_code == 200

    await app_client.delete(f"/collections/{item['collection']}/items/{item['id']}")


async def test_app_search_response(app_client, ctx):
    resp = await app_client.get("/search", params={"ids": ["test-item"]})
    assert resp.status_code == 200
    resp_json = resp.json()

    assert resp_json.get("type") == "FeatureCollection"
    # stac_version and stac_extensions were removed in v1.0.0-beta.3
    assert resp_json.get("stac_version") is None
    assert resp_json.get("stac_extensions") is None


async def test_app_context_extension(app_client, ctx, txn_client):
    test_item = ctx.item
    test_item["id"] = "test-item-2"
    test_item["collection"] = "test-collection-2"
    test_collection = ctx.collection
    test_collection["id"] = "test-collection-2"

    await create_collection(txn_client, test_collection)
    await create_item(txn_client, test_item)

    resp = await app_client.get(
        f"/collections/{test_collection['id']}/items/{test_item['id']}"
    )
    assert resp.status_code == 200
    resp_json = resp.json()
    assert resp_json["id"] == test_item["id"]
    assert resp_json["collection"] == test_item["collection"]

    resp = await app_client.get(f"/collections/{test_collection['id']}")
    assert resp.status_code == 200
    resp_json = resp.json()
    assert resp_json["id"] == test_collection["id"]

    resp = await app_client.post("/search", json={"collections": ["test-collection-2"]})
    assert resp.status_code == 200
    resp_json = resp.json()
    assert len(resp_json["features"]) == 1
    assert "context" in resp_json
    assert resp_json["context"]["returned"] == 1
    if matched := resp_json["context"].get("matched"):
        assert matched == 1


@pytest.mark.skip(reason="fields not implemented yet")
async def test_app_fields_extension(load_test_data, app_client, txn_client):
    item = load_test_data("test_item.json")
    txn_client.create_item(item, request=MockRequest, refresh=True)

    resp = await app_client.get("/search", params={"collections": ["test-collection"]})
    assert resp.status_code == 200
    resp_json = resp.json()
    assert list(resp_json["features"][0]["properties"]) == ["datetime"]

    txn_client.delete_item(item["id"], item["collection"])


@pytest.mark.skip(reason="Query not implemented")
async def test_app_query_extension_gt(app_client, ctx):
    params = {"query": {"proj:epsg": {"gt": ctx.item["properties"]["proj:epsg"]}}}
    resp = await app_client.post("/search", json=params)
    assert resp.status_code == 200
    resp_json = resp.json()
    assert len(resp_json["features"]) == 0


@pytest.mark.skip(reason="Query not implemented")
async def test_app_query_extension_gte(app_client, ctx):
    params = {"query": {"proj:epsg": {"gte": ctx.item["properties"]["proj:epsg"]}}}
    resp = await app_client.post("/search", json=params)

    assert resp.status_code == 200
    assert len(resp.json()["features"]) == 1


async def test_app_query_extension_limit_lt0(app_client):
    assert (await app_client.post("/search", json={"limit": -1})).status_code == 400


async def test_app_query_extension_limit_gt10000(app_client):
    assert (await app_client.post("/search", json={"limit": 10001})).status_code == 400


async def test_app_query_extension_limit_10000(app_client):
    params = {"limit": 10000}
    resp = await app_client.post("/search", json=params)
    assert resp.status_code == 200


@pytest.mark.skip(reason="Sort extension not implemented")
async def test_app_sort_extension(app_client, txn_client, ctx):
    first_item = ctx.item
    item_date = datetime.strptime(
        first_item["properties"]["datetime"], "%Y-%m-%dT%H:%M:%SZ"
    )

    second_item = dict(first_item)
    second_item["id"] = "another-item"
    another_item_date = item_date - timedelta(days=1)
    second_item["properties"]["datetime"] = another_item_date.strftime(
        "%Y-%m-%dT%H:%M:%SZ"
    )
    await create_item(txn_client, second_item)

    params = {
        "collections": [first_item["collection"]],
        "sortby": [{"field": "properties.datetime", "direction": "desc"}],
    }
    resp = await app_client.post("/search", json=params)
    assert resp.status_code == 200
    resp_json = resp.json()
    assert resp_json["features"][0]["id"] == first_item["id"]
    assert resp_json["features"][1]["id"] == second_item["id"]


async def test_search_invalid_date(app_client, ctx):
    params = {
        "datetime": "2020-XX-01/2020-10-30",
        "collections": [ctx.item["collection"]],
    }

    resp = await app_client.post("/search", json=params)
    assert resp.status_code == 400


async def test_search_point_intersects(app_client, ctx):
    point = [150.04, -33.14]
    intersects = {"type": "Point", "coordinates": point}

    params = {
        "intersects": intersects,
        "collections": [ctx.item["collection"]],
    }
    resp = await app_client.post("/search", json=params)

    assert resp.status_code == 200
    resp_json = resp.json()
    assert len(resp_json["features"]) == 1


@pytest.mark.skip(reason="Date search not implemented")
async def test_datetime_non_interval(app_client, ctx):
    dt_formats = [
        "2020-02-12T12:30:22+00:00",
        "2020-02-12T12:30:22.00Z",
        "2020-02-12T12:30:22Z",
        "2020-02-12T12:30:22.00+00:00",
    ]

    for dt in dt_formats:
        params = {
            "datetime": dt,
            "collections": [ctx.item["collection"]],
        }

        resp = await app_client.post("/search", json=params)
        assert resp.status_code == 200
        resp_json = resp.json()
        # datetime is returned in this format "2020-02-12T12:30:22Z"
        assert resp_json["features"][0]["properties"]["datetime"][0:19] == dt[0:19]


async def test_bbox_3d(app_client, ctx):
    australia_bbox = [106.343365, -47.199523, 0.1, 168.218365, -19.437288, 0.1]
    params = {
        "bbox": australia_bbox,
        # "collections": [ctx.item["collection"]],
    }
    resp = await app_client.post("/search", json=params)
    assert resp.status_code == 200
    resp_json = resp.json()
    assert len(resp_json["features"]) == 1


@pytest.mark.skip(reason="Line String inttersects not implemented")
async def test_search_line_string_intersects(app_client, ctx):
    line = [[150.04, -33.14], [150.22, -33.89]]
    intersects = {"type": "LineString", "coordinates": line}
    params = {
        "intersects": intersects,
        "collections": [ctx.item["collection"]],
    }

    resp = await app_client.post("/search", json=params)

    assert resp.status_code == 200

    resp_json = resp.json()
    assert len(resp_json["features"]) == 1


tile38.py
"""Test script for tile38."""
import asyncio
import json

from pyle38 import Tile38

# import os


tile38 = Tile38("redis://localhost:9851")


async def main():
    """Test script for tile38."""
    tile38 = Tile38(url="redis://localhost:9851", follower_url="redis://localhost:9851")
    item = {
        "type": "Feature",
        "id": "test-item-57",
        "stac_version": "1.0.0",
        "stac_extensions": [
            "https://stac-extensions.github.io/eo/v1.0.0/schema.json",
            "https://stac-extensions.github.io/projection/v1.0.0/schema.json",
        ],
        "geometry": {
            "coordinates": [
                [
                    [152.15052873427666, -33.82243006904891],
                    [150.1000346138806, -34.257132625788756],
                    [149.5776607193635, -32.514709769700254],
                    [151.6262528041627, -32.08081674221862],
                    [152.15052873427666, -33.82243006904891],
                ]
            ],
            "type": "Polygon",
        },
        "properties": {
            "datetime": "2018-02-12T12:30:22Z",
            "landsat:scene_id": "LC82081612020043LGN00",
            "landsat:row": "161",
            "gsd": 15,
            "landsat:revision": "00",
            "view:sun_azimuth": -148.83296771,
            "instrument": "OLI_TIRS",
            "landsat:product_id": "LC08_L1GT_208161_20200212_20200212_01_RT",
            "eo:cloud_cover": 0,
            "landsat:tier": "RT",
            "landsat:processing_level": "L1GT",
            "landsat:column": "208",
            "platform": "landsat-8",
            "proj:epsg": 32756,
            "view:sun_elevation": -37.30791534,
            "view:off_nadir": 0,
            "height": 2500,
            "width": 2500,
        },
        "bbox": [149.57574, -34.25796, 152.15194, -32.07915],
        "collection": "test-collection",
        "assets": {},
        "links": [
            {
                "href": "http://localhost:8081/collections/landsat-8-l1/items/LC82081612020043",
                "rel": "self",
                "type": "application/geo+json",
            },
            {
                "href": "http://localhost:8081/collections/landsat-8-l1",
                "rel": "parent",
                "type": "application/json",
            },
            {
                "href": "http://localhost:8081/collections/landsat-8-l1",
                "rel": "collection",
                "type": "application/json",
            },
            {
                "href": "http://localhost:8081/",
                "rel": "root",
                "type": "application/json",
            },
        ],
    }

    await tile38.set("test", item["id"]).object(item["geometry"]).exec()
    # await tile38.set("test", item["id"]).bounds(item["bbox"][0], item["bbox"][1], item["bbox"][2], item["bbox"][3]).exec()
    await tile38.jset("test", item["id"], "item", json.dumps(item))
    # bbox = [96.718129, -46.464978, 178.632191, -1.445461]
    # objects = await tile38.intersects("test").bounds(bbox[0], bbox[1], bbox[2], bbox[3]).asObjects()
    objects = await tile38.intersects("test").object(item["geometry"]).asObjects()
    items = []
    print(objects.count)
    # print(objects)
    # print(objects.objects[0].id)
    for i in range(objects.count):
        item_result = objects.objects[i].object
        item_result["id"] = objects.objects[i].id
        items.append(json.loads(item_result["item"]))
        # item = json.loads(objects.objects[i].object)
        # items.append(json.loads(item["item"]))
    # print(items)
    # print(type(items[0]))

    print(objects.objects[0].id)

    response = await tile38.get("test", objects.objects[0].id).asObject()
    print("-----------------")
    object = json.loads(response.object["item"])
    print(object)
    print(type(object))

    search = await tile38.search("test").match("test*").asStringObjects()
    print(search)

    # result = await tile38.get(item["collection"], item["id"]).asObject()
    # print(result.object)

    # print(type(result.object))

    # user = {
    #     "first": "Tom",
    #     "last": "Anderson",
    #     "dict": {
    #         "job": "janitor"
    #     }
    # }
    # await tile38.jset('user', 'Tom', 'user', json.dumps(user))
    # # await tile38.jset('user', 901, 'name.first', 'Anderson')
    # response = await tile38.get('user', 'Tom').asObject()

    # response = json.loads(response.object)
    # user = json.loads(response["user"])
    # print(user)
    # print(user["dict"]["job"])

    # print(type(user))
    # # print(response)

    # list = await tile38.scan('user').asObjects()
    # print(list)
    # print(list.count)
    # user = json.loads(list.objects[1].object)
    # print(json.loads(user["user"]))


asyncio.run(main())


stac_fastapi/caching/stac_fastapi/caching/core.py
"""Item crud client."""
import json
import logging
import time
from datetime import datetime as datetime_type
from datetime import timezone
from typing import List, Optional, Type, Union
from urllib.parse import urljoin

import attr
import stac_pydantic.api
from fastapi import HTTPException
from overrides import overrides
from pydantic import ValidationError
from stac_pydantic.links import Relations
from stac_pydantic.shared import MimeTypes
from starlette.requests import Request

from stac_fastapi.caching import serializers
from stac_fastapi.caching.config import Tile38Settings
from stac_fastapi.caching.database_logic import DatabaseLogic
from stac_fastapi.caching.models.links import PagingLinks
from stac_fastapi.caching.serializers import CollectionSerializer, ItemSerializer
from stac_fastapi.caching.session import Session
from stac_fastapi.extensions.third_party.bulk_transactions import (
    BaseBulkTransactionsClient,
    Items,
)
from stac_fastapi.types import stac as stac_types
from stac_fastapi.types.core import AsyncBaseCoreClient, AsyncBaseTransactionsClient
from stac_fastapi.types.links import CollectionLinks
from stac_fastapi.types.stac import Collection, Collections, Item, ItemCollection

logger = logging.getLogger(__name__)

NumType = Union[float, int]


@attr.s
class CoreClient(AsyncBaseCoreClient):
    """Client for core endpoints defined by stac."""

    session: Session = attr.ib(default=attr.Factory(Session.create_from_env))
    item_serializer: Type[serializers.ItemSerializer] = attr.ib(
        default=serializers.ItemSerializer
    )
    collection_serializer: Type[serializers.CollectionSerializer] = attr.ib(
        default=serializers.CollectionSerializer
    )
    database = DatabaseLogic()

    @overrides
    async def all_collections(self, **kwargs) -> Collections:
        """Read all collections from the database."""
        base_url = str(kwargs["request"].base_url)
        collection_list = await self.database.get_all_collections()
        collection_list = [
            self.collection_serializer.db_to_stac(c, base_url=base_url)
            for c in collection_list
        ]

        links = [
            {
                "rel": Relations.root.value,
                "type": MimeTypes.json,
                "href": base_url,
            },
            {
                "rel": Relations.parent.value,
                "type": MimeTypes.json,
                "href": base_url,
            },
            {
                "rel": Relations.self.value,
                "type": MimeTypes.json,
                "href": urljoin(base_url, "collections"),
            },
        ]

        return Collections(collections=collection_list, links=links)

    @overrides
    async def get_collection(self, collection_id: str, **kwargs) -> Collection:
        """Get collection by id."""
        base_url = str(kwargs["request"].base_url)
        collection = await self.database.find_collection(collection_id=collection_id)
        return self.collection_serializer.db_to_stac(collection, base_url)

    @overrides
    async def item_collection(
        self, collection_id: str, limit: int = 10, token: str = None, **kwargs
    ) -> ItemCollection:
        """Read an item collection from the database."""
        request: Request = kwargs["request"]
        base_url = str(kwargs["request"].base_url)

        items, maybe_count, next_token = await self.database.get_item_collection(
            collection_id=collection_id
        )

        items = [
            self.item_serializer.db_to_stac(item, base_url=base_url) for item in items
        ]

        context_obj = None
        if self.extension_is_enabled("ContextExtension"):
            context_obj = {
                "returned": len(items),
                "limit": limit,
            }
            if maybe_count is not None:
                context_obj["matched"] = maybe_count

        links = []
        if next_token:
            links = await PagingLinks(request=request, next=next_token).get_links()

        return ItemCollection(
            type="FeatureCollection",
            features=items,
            links=links,
            context=context_obj,
        )

    @overrides
    async def get_item(self, item_id: str, collection_id: str, **kwargs) -> Item:
        """Get item by item id, collection id."""
        base_url = str(kwargs["request"].base_url)
        item = await self.database.get_one_item(
            item_id=item_id, collection_id=collection_id
        )
        return self.item_serializer.db_to_stac(item, base_url)

    @staticmethod
    def _return_date(interval_str):
        intervals = interval_str.split("/")
        if len(intervals) == 1:
            datetime = intervals[0][0:19] + "Z"
            return {"eq": datetime}
        else:
            start_date = intervals[0]
            end_date = intervals[1]
            if ".." not in intervals:
                start_date = start_date[0:19] + "Z"
                end_date = end_date[0:19] + "Z"
            elif start_date != "..":
                start_date = start_date[0:19] + "Z"
                end_date = "2200-12-01T12:31:12Z"
            elif end_date != "..":
                start_date = "1900-10-01T00:00:00Z"
                end_date = end_date[0:19] + "Z"
            else:
                start_date = "1900-10-01T00:00:00Z"
                end_date = "2200-12-01T12:31:12Z"

            return {"lte": end_date, "gte": start_date}

    @overrides
    async def get_search(
        self,
        collections: Optional[List[str]] = None,
        ids: Optional[List[str]] = None,
        bbox: Optional[List[NumType]] = None,
        datetime: Optional[Union[str, datetime_type]] = None,
        limit: Optional[int] = 10,
        query: Optional[str] = None,
        token: Optional[str] = None,
        fields: Optional[List[str]] = None,
        sortby: Optional[str] = None,
        **kwargs,
    ) -> ItemCollection:
        """GET search catalog."""
        base_args = {
            "collections": collections,
            "ids": ids,
            "bbox": bbox,
            "limit": limit,
            "token": token,
            "query": json.loads(query) if query else query,
        }
        if datetime:
            base_args["datetime"] = datetime
        if sortby:
            # https://github.com/radiantearth/stac-spec/tree/master/api-spec/extensions/sort#http-get-or-post-form
            sort_param = []
            for sort in sortby:
                sort_param.append(
                    {
                        "field": sort[1:],
                        "direction": "asc" if sort[0] == "+" else "desc",
                    }
                )
            base_args["sortby"] = sort_param

        # if fields:
        #     includes = set()
        #     excludes = set()
        #     for field in fields:
        #         if field[0] == "-":
        #             excludes.add(field[1:])
        #         elif field[0] == "+":
        #             includes.add(field[1:])
        #         else:
        #             includes.add(field)
        #     base_args["fields"] = {"include": includes, "exclude": excludes}

        # Do the request
        try:
            search_request = self.post_request_model(**base_args)
        except ValidationError:
            raise HTTPException(status_code=400, detail="Invalid parameters provided")
        resp = await self.post_search(search_request, request=kwargs["request"])

        return resp

    @overrides
    async def post_search(
        self, search_request: stac_pydantic.api.Search, **kwargs
    ) -> ItemCollection:
        """POST search catalog."""
        request: Request = kwargs["request"]
        base_url = str(request.base_url)
        items = []
        count = 0

        if search_request.limit:
            limit = search_request.limit
        else:
            limit = 10

        if search_request.ids:
            items = await self.database.get_items(
                item_ids=search_request.ids
            )
            count = len(items)

        # if search_request.datetime:
        #     datetime_search = self._return_date(search_request.datetime)
        #     search = self.database.apply_datetime_filter(
        #         search=search, datetime_search=datetime_search
        #     )

        if search_request.bbox:
            bbox = search_request.bbox
            if len(bbox) == 6:
                bbox = [bbox[0], bbox[1], bbox[3], bbox[4]]
            items, count = await self.database.apply_bbox_filter(
                collection_id="test-collection", bbox=bbox, limit=limit
            )

        if search_request.intersects:
            items, count = await self.database.apply_intersects_filter(
                intersects=search_request.intersects, limit=limit
            )

        # if search_request.query:
        #     for (field_name, expr) in search_request.query.items():
        #         field = "properties__" + field_name
        #         for (op, value) in expr.items():
        #             search = self.database.apply_stacql_filter(
        #                 search=search, op=op, field=field, value=value
        #             )

        # sort = None
        # if search_request.sortby:
        #     sort = self.database.populate_sort(search_request.sortby)

        if search_request.collections:
            if len(items) > 0:
                items_copy = items
                items = []
                count = 0
                for item in items_copy:
                    if item["collection"] in search_request.collections:
                        items.append(item)
                        count += 1
            else:
                items = []
                for collection_id in search_request.collections:
                    items_list, count, _ = await self.database.get_item_collection(
                        collection_id=collection_id
                    )
                    items = items + items_list

        items = [
            self.item_serializer.db_to_stac(item, base_url=base_url) for item in items
        ]

        context_obj = None
        if self.extension_is_enabled("ContextExtension"):
            context_obj = {
                "returned": len(items),
                "limit": limit,
            }
            if count is not None:
                context_obj["matched"] = count

        links = []
        # if next_token:
        #     links = await PagingLinks(request=request, next=next_token).get_links()

        return ItemCollection(
            type="FeatureCollection",
            features=items,
            links=links,
            context=context_obj,
        )


@attr.s
class TransactionsClient(AsyncBaseTransactionsClient):
    """Transactions extension specific CRUD operations."""

    session: Session = attr.ib(default=attr.Factory(Session.create_from_env))
    database = DatabaseLogic()

    @overrides
    async def create_item(self, item: stac_types.Item, **kwargs) -> stac_types.Item:
        """Create item."""
        base_url = str(kwargs["request"].base_url)

        # If a feature collection is posted
        if item["type"] == "FeatureCollection":
            bulk_client = BulkTransactionsClient()
            processed_items = [
                bulk_client.preprocess_item(item, base_url) for item in item["features"]  # type: ignore
            ]
            await self.database.bulk_async(
                processed_items, refresh=kwargs.get("refresh", False)
            )

            return None  # type: ignore
        else:
            item = await self.database.prep_create_item(item=item, base_url=base_url)
            await self.database.create_item(item, refresh=kwargs.get("refresh", False))
            return item

    @overrides
    async def update_item(self, item: stac_types.Item, **kwargs) -> stac_types.Item:
        """Update item."""
        base_url = str(kwargs["request"].base_url)
        now = datetime_type.now(timezone.utc).isoformat().replace("+00:00", "Z")
        item["properties"]["updated"] = str(now)

        await self.database.check_collection_exists(collection_id=item["collection"])
        # todo: index instead of delete and create
        await self.delete_item(item_id=item["id"], collection_id=item["collection"])
        time.sleep(1)
        await self.create_item(item=item, **kwargs)

        return ItemSerializer.db_to_stac(item, base_url)

    @overrides
    async def delete_item(
        self, item_id: str, collection_id: str, **kwargs
    ) -> stac_types.Item:
        """Delete item."""
        await self.database.delete_item(item_id=item_id, collection_id=collection_id)
        return None  # type: ignore

    @overrides
    async def create_collection(
        self, collection: stac_types.Collection, **kwargs
    ) -> stac_types.Collection:
        """Create collection."""
        base_url = str(kwargs["request"].base_url)
        collection_links = CollectionLinks(
            collection_id=collection["id"], base_url=base_url
        ).create_links()
        collection["links"] = collection_links
        await self.database.create_collection(collection=collection)

        return CollectionSerializer.db_to_stac(collection, base_url)

    @overrides
    async def update_collection(
        self, collection: stac_types.Collection, **kwargs
    ) -> stac_types.Collection:
        """Update collection."""
        base_url = str(kwargs["request"].base_url)

        await self.database.find_collection(collection_id=collection["id"])
        await self.delete_collection(collection["id"])
        time.sleep(1)
        await self.create_collection(collection, **kwargs)

        return CollectionSerializer.db_to_stac(collection, base_url)

    @overrides
    async def delete_collection(
        self, collection_id: str, **kwargs
    ) -> stac_types.Collection:
        """Delete collection."""
        await self.database.delete_collection(collection_id=collection_id)
        return None  # type: ignore


@attr.s
class BulkTransactionsClient(BaseBulkTransactionsClient):
    """Postgres bulk transactions."""

    session: Session = attr.ib(default=attr.Factory(Session.create_from_env))
    database = DatabaseLogic()

    def __attrs_post_init__(self):
        """Create es engine."""
        settings = Tile38Settings()
        self.client = settings.create_client

    def preprocess_item(self, item: stac_types.Item, base_url) -> stac_types.Item:
        """Preprocess items to match data model."""
        return self.database.sync_prep_create_item(item=item, base_url=base_url)

    @overrides
    def bulk_item_insert(
        self, items: Items, chunk_size: Optional[int] = None, **kwargs
    ) -> str:
        """Bulk item insertion using es."""
        request = kwargs.get("request")
        if request:
            base_url = str(request.base_url)
        else:
            base_url = ""

        processed_items = [
            self.preprocess_item(item, base_url) for item in items.items.values()
        ]

        self.database.bulk_sync(processed_items, refresh=kwargs.get("refresh", False))

        return f"Successfully added {len(processed_items)} Items."
